### DISCUSSION TITLE
Development Processes

### DISCUSSION THREAD

**Discussion by frisitano** [scroll-tech/reth#3] [General]
# Overview
Our ambitions with this project are to implement a scroll node using reth in an SDK like manner. As reth may not provide the abstractions we require to achieve an SDK like implementation from day one we may find that we have to introduce features on this fork of reth. However, we want these features to be short lived and our ambition is to push the required abstractions upstream such that we can remove the diffs in a timely fashion.

## Branching Strategy
The branching strategy we will adopt is intended to minimise the diff and maintain in sync with upstream as much as possible. The branching strategy will only cover the development phase of the project, as we approach our first major version release we should update the strategy to support releases.

In our branching strategy we will introduce a new branch `scroll`. This branch will act as the main branch from which features will be developed. As a feature development begins a new feature branch will be created from the `scroll` branch. The `main` branch will be kept in sync with `main` from upstream and we will periodically merge `main` into `scroll`. We must define the triggers and frequency in which we merge upstream into `scroll`. A simple proposal would be to sync at least every 2 weeks. If we have to merge before the 2 week threshold then the timer will reset at the time of the merge.

## Development Process
In the circumstance that we need to introduce a new feature to reth which requires us to diverge from upstream we should develop and abstraction that allows us to remove the diff before we move onto the next feature.

**Comment by Thegaram** [comment]
Added a diffview page so that we can keep an eye on the fork diff size: https://scroll-tech.github.io/reth

> we should develop and abstraction

It will be a bit tricky to then submit this abstraction back to upstream, since the `scroll` branch will have these new abstractions interleaved with other changes. But it's probably acceptable, we just need to cherry-pick relevant changes to a new branch and issue a PR to reth from that branch.

**Comment by Thegaram** [comment]
Here is the current process to sync with upstream:

```
> git checkout main
> git pull origin main

> git fetch upstream
> git pull upstream main
> git push origin main

> COMMIT=$(git log -1 --pretty=format:%H)
> DIGEST=$(git log -1 --pretty=format:%h)

> git checkout scroll
> git pull origin scroll

> git checkout -b "chore-merge-upstream-$DIGEST"
> git merge main

> sed -i -e "s/hash: .*/hash: $COMMIT/g" fork.yaml
> git add fork.yaml
> git commit -m "update fork base commit"

> git push origin "chore-merge-upstream-$DIGEST"
```
